File: 백준/Silver/1189. 컴백홈/컴백홈.cc
#include <bits/stdc++.h>

using namespace std;

int n, m, k, visited[10][10];
char a[10][10];
const int dy[] = {-1,0,1,0};
const int dx[] = {0,1,0,-1};

int go(int y, int x) {
    if(y == 0 && x == m - 1) {
        if(k == visited[y][x]) return 1;
        return 0;
    }
    int ret = 0;
    for(int i = 0; i < 4; i++) {
        int ny = y + dy[i];
        int nx = x + dx[i];
        if(ny < 0 || ny >= n || nx < 0 || nx >= m || visited[ny][nx]) continue;
        if(a[ny][nx] == 'T') continue;
        visited[ny][nx] = visited[y][x] + 1;
        ret += go(ny, nx);
        visited[ny][nx] = 0;
    }
    return ret;
}

int main() {
    cin >> n >> m >> k;
    for(int i = 0; i < n; i++) {
        for(int j = 0; j < m; j++) {
            cin >> a[i][j];
        }
    }
    
    visited[n - 1][0] = 1;
    
    cout << go(n - 1, 0) << '\n';
    
    return 0;
}

File: 백준/Silver/1189. 컴백홈/컴백홈.swift
let input = readLine()!.split(separator: " ").map { Int($0)! }
let (r, c, k) = (input[0], input[1], input[2])
var arr: [[Character]] = []
for _ in 0..<r {
    let line = Array(readLine()!)
    arr.append(line)
} 

var visited = Array(repeating: Array(repeating: 0, count: c), count: r)

let dy = [-1, 0, 1, 0]
let dx = [0, 1, 0, -1]

func go(_ y: Int, _ x: Int) -> Int {
    if y == 0 && x == c - 1 {
        if visited[y][x] == k {
            return 1
        } else {
            return 0
        }
    }

    if visited[y][x] >= k { return 0 }
    var ret = 0
    for i in 0..<4 {
        let ny = y + dy[i], nx = x + dx[i]

        if ny < 0 || ny >= r || nx < 0 || nx >= c { continue }
        if visited[ny][nx] != 0 || arr[ny][nx] == "T" { continue }
        visited[ny][nx] = visited[y][x] + 1
        ret += go(ny, nx)
        visited[ny][nx] = 0
    }

    return ret
}

visited[r - 1][0] = 1
let res = go(r - 1, 0)

print(res)
